### ISSUE TITLE
Torrent Resuming Fails

### ISSUE THREAD

**Issue by MyNESCreations** [alanmcgovern/monotorrent#541]
I have it using the default setting of AutoSaveLoadFastResume (Also manually setting it to true) but when I start the torrent again, it fails instead of continuing from where it left off.
When I Add it it shows the correct data that it would if it loaded the fastresume, the progress is corrent and everything, but as soon as you call StartAsync() on it, or call StartAllAsync() on the engine, then it switches to downloading, resets the progress, and then after a second or so it switches to Failed state.

When stopping the applications I am calling;
     
        await clientEngine.StopAllAsync();

How I am resuming torrents;

        async void ResumeTorrents()
        {
            var metadataDinfo = new DirectoryInfo(clientEngine.Settings.MetadataCacheDirectory);
            if (metadataDinfo.Exists)
                foreach (var finfo in metadataDinfo.GetFiles("*.torrent"))
                {
                    TorrentDataList.Add(new TorrentData(finfo.Name, "", 0, "", "", "", "", "", ""));
                    await clientEngine.AddAsync(await Torrent.LoadAsync(finfo.FullName), settings.defaultTorrentSavePath);
                }
        }

I also al using AvaloniaUI, I don't know if there is any issues with that, but no exceptions or errors have been thrown from what I have seen.
I tried looking for more documentation but it seems most of it is either outdated or I am having trouble finding something.
I assume I am just missing something but was not sure so I wanted to check, any help is appreciated thanks!

**Comment by alanmcgovern** [comment]
Which version of the library are you using? The latest 2.x stable?

**Comment by alanmcgovern** [comment]
If you're using a build from master, or a 3.x pre-release, could you tell me if the torrent is v1, v2, or a hybrid one? You can check by looking at:

```
var t = await Torrent.LoadAsync(finfo.FullName);
if (t.InfoHashes.IsHybrid)
  // hybrid v1/v2
else if (t.InfoHashes.V2 != null)
  // v2 only
else
  // v1 only
```

**Comment by MyNESCreations** [comment]
Yes I am using the latest release 2.05.

**Comment by alanmcgovern** [comment]
> then after a second or so it switches to Failed state

Hrmm, this is the part that's most interesting. When this occurs, is there anything interesting in the `TorrentManager.Error` property? If it's non-null then it might have have the stacktrace detailing the underlying failure.

If I had to guess, I'd say there's an unexpected file sharing issue. If that property is null I'll probably need more info to dig into the underlying issue. Does it happen for all torrents, or just one specifically?

**Comment by MyNESCreations** [comment]
Okay yeah thank you I just took a loot at what `TorrentManager.Error` had in it, and it is saying that it is a Write Failure.
I realize I should probably also note I am on Linux, I am not sure what that changes but probably also important.

The main torrent I have been testing with is the Arch Linux iso, but I have also tried a few others and they all are behaving the same way.

`"Access to the path '/mnt/Projectz/Torrents/Torrent Files/Other/archlinux-2022.06.01-x86_64.iso' is denied."`

Stack Trace;
```at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at MonoTorrent.Client.TorrentFileStream..ctor(String path, FileAccess access)
   at MonoTorrent.Client.PieceWriters.DiskWriter.<>c.<.cctor>b__21_0(ITorrentFileInfo file, FileAccess access)
   at MonoTorrent.Client.PieceWriters.FileStreamBuffer.GetOrCreateStreamAsync(ITorrentFileInfo file, FileAccess access)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter`1.GetResult()
   at MonoTorrent.Client.PieceWriters.DiskWriter.WriteAsync(ITorrentFileInfo file, Int64 offset, Byte[] buffer, Int32 bufferOffset, Int32 count)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter.GetResult()
   at MonoTorrent.Client.PieceWriters.IPieceWriterExtensions.WriteToFilesAsync(IPieceWriter writer, ITorrentData manager, BlockInfo request, Byte[] buffer)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter.GetResult()
   at MonoTorrent.Client.PieceWriters.MemoryCache.WriteAsync(ITorrentData torrent, BlockInfo block, Byte[] buffer, Boolean preferSkipCache)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter.GetResult()
   at MonoTorrent.Client.DiskManager.WriteAsync(ITorrentData manager, BlockInfo request, Byte[] buffer)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter.GetResult()
   at MonoTorrent.Client.Modes.Mode.WritePieceAsync(PieceMessage message, IList`1 peersInvolved)```

**Comment by alanmcgovern** [comment]
Hrm, this is a bit strange. 

Can you confirm there are no other processes running which are accessing those files? MonoTorrent opens files with FileShare.ReadWrite ( https://github.com/alanmcgovern/monotorrent/blob/9095e775ab06c16a83d7a5f35298fc8c6a9725b2/src/MonoTorrent.PieceWriter/MonoTorrent.PieceWriter/DiskWriter.cs#L47 ) so it should be able to access these files as long as other processes which open them have not used more restrictive share options, and also the OS supports this capability. 

Linux has issues with locking ( this is still authoritative I think https://github.com/dotnet/runtime/issues/24432 ). I assume this means those options will 'never' work on Linux as expected, and may just need to be removed. 

OOC, if you are able to trigger this issue where MonoTorrent is the only program accessing these files, I'm interested in knowing how.  If you are certain there's no stale processes hanging around holding the files open, then there's an oops somewhere which should be resolvable.

**Comment by MyNESCreations** [comment]
Well thank you so much for all the help but it turns out, as it often does haha, the issue was just something I was doing in my own code.
I had wanted all torrents to be in a containing directory even single file torrents, so I was doing it manually for all torrents so I set `TorrentSettings.CreateContainingDirectory` to false, and I was using the name of the torrent as the name of the directory to download into. The reason it worked perfectly the first time is because I was doing;
`Path.Combine(settings.defaultTorrentSavePath, torrent.Name)` but later when I was loading the torrent I forgot to use the same way of building the path, and instead was just using `settings.defaultTorrentSavePath`. So it was trying to resume the torrent thinking the directory was the file.
And the reason this was still failing for each torrent I tried is because the name of the torrent was the same as the name of the main file, so even if there was multiple, at least one of them was named the same as the directory so it failed.

So I fixed it to use `Path.Combine(settings.defaultTorrentSavePath, torrent.Name)` in both locations and it works perfectly!
Thank you again so much for the help!
